File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

type MessageType byte

const (
	BetMsg  MessageType = 1
	Finish  MessageType = 2
	Winners MessageType = 3
)

type Bet struct {
	AgencyId  int
	FirstName string
	LastName  string
	Document  int
	Birthdate string
	Number    int
}

type BetMessage struct {
	Bets []Bet
}

type FinishMessage struct{}

type WinnersMessage struct {
	Winners []Bet
}

func encodeString(value string) []byte {
	raw := []byte(value)
	buf := make([]byte, 2+len(raw))
	binary.BigEndian.PutUint16(buf[:2], uint16(len(raw)))
	copy(buf[2:], raw)
	return buf
}

func decodeString(data []byte, offset int) (string, int, error) {
	if offset+2 > len(data) {
		return "", 0, fmt.Errorf("truncated string length")
	}
	length := int(binary.BigEndian.Uint16(data[offset : offset+2]))
	offset += 2
	if offset+length > len(data) {
		return "", 0, fmt.Errorf("truncated string data")
	}
	return string(data[offset : offset+length]), offset + length, nil
}

func serializeBet(bet Bet) []byte {
	payload := make([]byte, 0, 32)
	var buf [4]byte

	binary.BigEndian.PutUint32(buf[:], uint32(bet.AgencyId))
	payload = append(payload, buf[:]...)
	payload = append(payload, encodeString(bet.FirstName)...)
	payload = append(payload, encodeString(bet.LastName)...)
	binary.BigEndian.PutUint32(buf[:], uint32(bet.Document))
	payload = append(payload, buf[:]...)
	payload = append(payload, encodeString(bet.Birthdate)...)
	binary.BigEndian.PutUint32(buf[:], uint32(bet.Number))
	payload = append(payload, buf[:]...)
	return payload
}

func deserializeBet(data []byte, offset int) (Bet, int, error) {
	if offset+4 > len(data) {
		return Bet{}, 0, fmt.Errorf("truncated agency id")
	}
	agencyID := int(int32(binary.BigEndian.Uint32(data[offset : offset+4])))
	offset += 4

	firstName, next, err := decodeString(data, offset)
	if err != nil {
		return Bet{}, 0, fmt.Errorf("decode first_name: %w", err)
	}
	offset = next

	lastName, next, err := decodeString(data, offset)
	if err != nil {
		return Bet{}, 0, fmt.Errorf("decode last_name: %w", err)
	}
	offset = next

	if offset+4 > len(data) {
		return Bet{}, 0, fmt.Errorf("truncated document")
	}
	document := int(int32(binary.BigEndian.Uint32(data[offset : offset+4])))
	offset += 4

	birthdate, next, err := decodeString(data, offset)
	if err != nil {
		return Bet{}, 0, fmt.Errorf("decode birthdate: %w", err)
	}
	offset = next

	if offset+4 > len(data) {
		return Bet{}, 0, fmt.Errorf("truncated number")
	}
	number := int(int32(binary.BigEndian.Uint32(data[offset : offset+4])))
	offset += 4

	return Bet{
		AgencyId:  agencyID,
		FirstName: firstName,
		LastName:  lastName,
		Document:  document,
		Birthdate: birthdate,
		Number:    number,
	}, offset, nil
}

func serializeBets(bets []Bet) []byte {
	payload := make([]byte, 4)
	binary.BigEndian.PutUint32(payload, uint32(len(bets)))
	for _, bet := range bets {
		payload = append(payload, serializeBet(bet)...)
	}
	return payload
}

func deserializeBets(data []byte, offset int) ([]Bet, int, error) {
	if offset+4 > len(data) {
		return nil, 0, fmt.Errorf("truncated bet count")
	}
	count := int(binary.BigEndian.Uint32(data[offset : offset+4]))
	offset += 4

	bets := make([]Bet, 0, count)
	for i := 0; i < count; i++ {
		bet, next, err := deserializeBet(data, offset)
		if err != nil {
			return nil, 0, fmt.Errorf("decode bet[%d]: %w", i, err)
		}
		bets = append(bets, bet)
		offset = next
	}
	return bets, offset, nil
}

func SerializeMessage(msgType MessageType, payload interface{}) ([]byte, error) {
	var payloadBytes []byte

	switch msgType {
	case BetMsg:
		betMsg, ok := payload.(BetMessage)
		if !ok {
			return nil, fmt.Errorf("invalid BET payload")
		}
		payloadBytes = serializeBets(betMsg.Bets)
	case Winners:
		winnersMsg, ok := payload.(WinnersMessage)
		if !ok {
			return nil, fmt.Errorf("invalid WINNERS payload")
		}
		payloadBytes = serializeBets(winnersMsg.Winners)
	case Finish:
		payloadBytes = nil
	default:
		return nil, fmt.Errorf("unknown message type: %v", msgType)
	}

	frame := make([]byte, 5+len(payloadBytes))
	frame[0] = byte(msgType)
	binary.BigEndian.PutUint32(frame[1:5], uint32(len(payloadBytes)))
	copy(frame[5:], payloadBytes)
	return frame, nil
}

func RecvMessage(socket io.Reader) (MessageType, interface{}, error) {
	header, err := safe_socket.RecvAll(socket, 5)
	if err != nil {
		return 0, nil, fmt.Errorf("failed to read message header: %w", err)
	}

	msgType := MessageType(header[0])
	payloadLength := int(binary.BigEndian.Uint32(header[1:5]))
	payloadBytes, err := safe_socket.RecvAll(socket, payloadLength)
	if err != nil {
		return 0, nil, fmt.Errorf("failed to read message payload: %w", err)
	}

	switch msgType {
	case BetMsg:
		bets, _, err := deserializeBets(payloadBytes, 0)
		if err != nil {
			return 0, nil, fmt.Errorf("failed to deserialize bets: %w", err)
		}
		return BetMsg, BetMessage{Bets: bets}, nil
	case Finish:
		return Finish, FinishMessage{}, nil
	case Winners:
		winners, _, err := deserializeBets(payloadBytes, 0)
		if err != nil {
			return 0, nil, fmt.Errorf("failed to deserialize winners: %w", err)
		}
		return Winners, WinnersMessage{Winners: winners}, nil
	default:
		return 0, nil, fmt.Errorf("unknown message type: %v", msgType)
	}
}

func SendMessage(socket io.Writer, msgType MessageType, payload interface{}) error {
	messageBytes, err := SerializeMessage(msgType, payload)
	if err != nil {
		return err
	}
	return safe_socket.SendAll(socket, messageBytes)
}

File: services/server/src/protocol/protocol.py
import struct
from dataclasses import dataclass
from typing import List

import safe_socket


@dataclass
class Bet:
    agency_id: int
    first_name: str
    last_name: str
    document: int
    birthdate: str
    number: int


@dataclass
class BetMessage:
    bets: List[Bet]


@dataclass
class FinishMessage:
    pass


@dataclass
class WinnersMessage:
    winners: List[Bet]


class MessageType:
    BET = 1
    FINISH = 2
    WINNERS = 3


def _encode_string(value: str) -> bytes:
    encoded = value.encode("utf-8")
    return struct.pack(">H", len(encoded)) + encoded


def _decode_string(payload: bytes, offset: int):
    if offset + 2 > len(payload):
        raise ValueError("truncated string length")
    length = struct.unpack(">H", payload[offset : offset + 2])[0]
    offset += 2
    if offset + length > len(payload):
        raise ValueError("truncated string payload")
    value = payload[offset : offset + length].decode("utf-8")
    return value, offset + length


def serialize_bet(bet: Bet) -> bytes:
    payload = bytearray()
    payload.extend(struct.pack(">i", bet.agency_id))
    payload.extend(_encode_string(bet.first_name))
    payload.extend(_encode_string(bet.last_name))
    payload.extend(struct.pack(">i", bet.document))
    payload.extend(_encode_string(bet.birthdate))
    payload.extend(struct.pack(">i", bet.number))
    return bytes(payload)


def deserialize_bet(payload: bytes, offset: int = 0):
    if offset + 4 > len(payload):
        raise ValueError("truncated bet header")
    agency_id = struct.unpack(">i", payload[offset : offset + 4])[0]
    offset += 4

    first_name, offset = _decode_string(payload, offset)
    last_name, offset = _decode_string(payload, offset)

    if offset + 4 > len(payload):
        raise ValueError("truncated bet document")
    document = struct.unpack(">i", payload[offset : offset + 4])[0]
    offset += 4

    birthdate, offset = _decode_string(payload, offset)

    if offset + 4 > len(payload):
        raise ValueError("truncated bet number")
    number = struct.unpack(">i", payload[offset : offset + 4])[0]
    offset += 4

    return Bet(
        agency_id=agency_id,
        first_name=first_name,
        last_name=last_name,
        document=document,
        birthdate=birthdate,
        number=number,
    ), offset


def serialize_bets(bets: List[Bet]) -> bytes:
    payload = bytearray()
    payload.extend(struct.pack(">I", len(bets)))
    for bet in bets:
        payload.extend(serialize_bet(bet))
    return bytes(payload)


def deserialize_bets(payload: bytes, offset: int = 0):
    if offset + 4 > len(payload):
        raise ValueError("truncated bets count")
    bet_count = struct.unpack(">I", payload[offset : offset + 4])[0]
    offset += 4

    bets = []
    for _ in range(bet_count):
        bet, offset = deserialize_bet(payload, offset)
        bets.append(bet)
    return bets, offset


def serialize_message(msg_type: int, payload) -> bytes:
    if msg_type == MessageType.BET:
        payload_bytes = serialize_bets(payload.bets)
    elif msg_type == MessageType.WINNERS:
        payload_bytes = serialize_bets(payload.winners)
    elif msg_type == MessageType.FINISH:
        payload_bytes = b""
    else:
        raise ValueError(f"Unknown message type: {msg_type}")

    return struct.pack(">B", msg_type) + struct.pack(">I", len(payload_bytes)) + payload_bytes


def recv_message(socket) -> tuple:
    header = safe_socket.recv_all(socket, 5)
    msg_type = header[0]
    payload_length = struct.unpack(">I", header[1:5])[0]
    payload_bytes = safe_socket.recv_all(socket, payload_length)

    if msg_type == MessageType.BET:
        bets, _ = deserialize_bets(payload_bytes)
        return msg_type, BetMessage(bets=bets)
    if msg_type == MessageType.FINISH:
        return msg_type, FinishMessage()
    if msg_type == MessageType.WINNERS:
        winners, _ = deserialize_bets(payload_bytes)
        return msg_type, WinnersMessage(winners=winners)

    raise ValueError(f"Unknown message type: {msg_type}")


def send_message(socket, msg_type: int, payload) -> None:
    message_bytes = serialize_message(msg_type, payload)
    safe_socket.send_all(socket, message_bytes)
